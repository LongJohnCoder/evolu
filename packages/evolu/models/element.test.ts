import { createStableIDFactory } from '../../web/tests/integration/helpers';
import {
  deleteContentElement,
  Element,
  isNormalizedElement,
  materializePath,
  normalizeElement,
} from './element';

const id = createStableIDFactory();

test('normalizeElement merges adjacent texts', () => {
  const element: Element = {
    id: id(),
    children: [
      { id: id(), text: 'a' },
      { id: id(), text: 'b' },
      {
        id: id(),
        children: [
          { id: id(), text: 'a' },
          { id: id(), text: 'b' },
          { id: id(), text: 'c' },
          {
            id: id(),
            children: [
              { id: id(), text: 'a' },
              { id: id(), children: [] },
              { id: id(), text: 'b' },
              { id: id(), text: 'c' },
            ],
          },
        ],
      },
      { id: id(), text: 'a' },
      { id: id(), text: '' },
      { id: id(), text: 'b' },
    ],
  };
  expect(normalizeElement(element)).toMatchSnapshot();
});

test('normalizeElement preserves identity', () => {
  const element1: Element = {
    id: id(),
    children: [],
  };
  expect(normalizeElement(element1)).toBe(element1);

  const element2: Element = {
    id: id(),
    children: [{ id: id(), children: [] }, { id: id(), text: 'a' }],
  };
  expect(normalizeElement(element2)).toBe(element2);

  const preservedChild = { id: id(), children: [] };
  const element3: Element = {
    id: id(),
    children: [
      preservedChild,
      { id: id(), text: 'a' },
      { id: id(), text: 'b' },
    ],
  };
  const normalizedElement3 = normalizeElement(element3);
  expect(normalizedElement3).not.toBe(element3);
  expect(normalizedElement3.children[0]).toBe(preservedChild);
});

test('isNormalizedElement', () => {
  expect(
    isNormalizedElement({
      id: id(),
      children: [{ id: id(), text: 'a' }],
    }),
  ).toBe(true);

  // Empty string is BR, that's ok.
  expect(
    isNormalizedElement({
      id: id(),
      children: [{ id: id(), text: '' }],
    }),
  ).toBe(true);

  // Two not empty string, that's not ok.
  expect(
    isNormalizedElement({
      id: id(),
      children: [{ id: id(), text: 'a' }, { id: id(), text: 'b' }],
    }),
  ).toBe(false);

  // Recursion works.
  expect(
    isNormalizedElement({
      id: id(),
      children: [{ id: id(), children: [{ id: id(), text: '' }] }],
    }),
  ).toBe(true);

  // Empty string is BR, so it's ok.
  expect(
    isNormalizedElement({
      id: id(),
      children: [
        { id: id(), text: 'a' },
        { id: id(), text: '' },
        { id: id(), text: 'a' },
      ],
    }),
  ).toBe(true);
});

test('normalizeElement do not add children', () => {
  expect(normalizeElement({ id: id(), children: [] })).toMatchSnapshot();
});

test('normalizeElement do not remove children', () => {
  expect(normalizeElement({ id: id(), children: [] })).toMatchSnapshot();
  expect(
    normalizeElement({ id: id(), children: [{ id: id(), text: '' }] }),
  ).toMatchSnapshot();
  expect(
    normalizeElement({ id: id(), children: [{ id: id(), text: '.' }] }),
  ).toMatchSnapshot();
});

test('deleteContentElement', () => {
  const el = { id: id(), children: [{ id: id(), text: 'a' }] };
  expect(
    deleteContentElement({ start: [0, 0], end: [0, 1] })(el),
  ).toMatchSnapshot();
});

test('materializePath', () => {
  // <div><b>a</b></div>
  const text = { id: id(), text: 'a' };
  const b = { id: id(), children: [text] };
  const div = { id: id(), children: [b] };

  expect(materializePath([])(div)).toMatchSnapshot();
  expect(materializePath([0])(div)).toMatchSnapshot();
  expect(materializePath([0, 0])(div)).toMatchSnapshot();
  expect(materializePath([0, 0, 0])(div)).toMatchSnapshot();
  expect(materializePath([0, 0, 1])(div)).toMatchSnapshot();

  // Nulls.
  expect(materializePath([0, 0, 0, 0])(div)).toMatchSnapshot();
  expect(materializePath([1])(div)).toMatchSnapshot();
  expect(materializePath([0, 1])(div)).toMatchSnapshot();
  expect(materializePath([0, 0, 2])(div)).toMatchSnapshot();
});
