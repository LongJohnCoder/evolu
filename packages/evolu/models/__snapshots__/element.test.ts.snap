// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`deleteContentElement 1`] = `
Object {
  "children": Array [
    Object {
      "id": "1",
      "text": "",
    },
  ],
  "id": "0",
}
`;

exports[`materializeEditorPath 1`] = `
Object {
  "_tag": "Some",
  "value": Object {
    "parents": Array [],
    "to": Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "id": "1",
              "text": "a",
            },
          ],
          "id": "2",
        },
      ],
      "id": "3",
    },
  },
}
`;

exports[`materializeEditorPath 2`] = `
Object {
  "_tag": "Some",
  "value": Object {
    "parents": Array [
      Object {
        "children": Array [
          Object {
            "children": Array [
              Object {
                "id": "1",
                "text": "a",
              },
            ],
            "id": "2",
          },
        ],
        "id": "3",
      },
    ],
    "to": Object {
      "children": Array [
        Object {
          "id": "1",
          "text": "a",
        },
      ],
      "id": "2",
    },
  },
}
`;

exports[`materializeEditorPath 3`] = `
Object {
  "_tag": "Some",
  "value": Object {
    "parents": Array [
      Object {
        "children": Array [
          Object {
            "children": Array [
              Object {
                "id": "1",
                "text": "a",
              },
            ],
            "id": "2",
          },
        ],
        "id": "3",
      },
      Object {
        "children": Array [
          Object {
            "id": "1",
            "text": "a",
          },
        ],
        "id": "2",
      },
    ],
    "to": Object {
      "id": "1",
      "text": "a",
    },
  },
}
`;

exports[`materializeEditorPath 4`] = `
Object {
  "_tag": "Some",
  "value": Object {
    "parents": Array [
      Object {
        "children": Array [
          Object {
            "children": Array [
              Object {
                "id": "1",
                "text": "a",
              },
            ],
            "id": "2",
          },
        ],
        "id": "3",
      },
      Object {
        "children": Array [
          Object {
            "id": "1",
            "text": "a",
          },
        ],
        "id": "2",
      },
    ],
    "to": Object {
      "editorText": Object {
        "id": "1",
        "text": "a",
      },
      "offset": 0,
    },
  },
}
`;

exports[`materializeEditorPath 5`] = `
Object {
  "_tag": "Some",
  "value": Object {
    "parents": Array [
      Object {
        "children": Array [
          Object {
            "children": Array [
              Object {
                "id": "1",
                "text": "a",
              },
            ],
            "id": "2",
          },
        ],
        "id": "3",
      },
      Object {
        "children": Array [
          Object {
            "id": "1",
            "text": "a",
          },
        ],
        "id": "2",
      },
    ],
    "to": Object {
      "editorText": Object {
        "id": "1",
        "text": "a",
      },
      "offset": 1,
    },
  },
}
`;

exports[`materializeEditorPath 6`] = `
Object {
  "_tag": "None",
}
`;

exports[`materializeEditorPath 7`] = `
Object {
  "_tag": "None",
}
`;

exports[`materializeEditorPath 8`] = `
Object {
  "_tag": "None",
}
`;

exports[`materializeEditorPath 9`] = `
Object {
  "_tag": "None",
}
`;

exports[`normalizeEditorElement do not add children 1`] = `
Object {
  "children": Array [],
  "id": "0",
}
`;

exports[`normalizeEditorElement do not remove children 1`] = `
Object {
  "children": Array [],
  "id": "0",
}
`;

exports[`normalizeEditorElement do not remove children 2`] = `
Object {
  "children": Array [
    Object {
      "id": "2",
      "text": "",
    },
  ],
  "id": "1",
}
`;

exports[`normalizeEditorElement do not remove children 3`] = `
Object {
  "children": Array [
    Object {
      "id": "4",
      "text": ".",
    },
  ],
  "id": "3",
}
`;

exports[`normalizeEditorElement merges adjacent texts 1`] = `
Object {
  "children": Array [
    Object {
      "id": "1",
      "text": "ab",
    },
    Object {
      "children": Array [
        Object {
          "id": "4",
          "text": "abc",
        },
        Object {
          "children": Array [
            Object {
              "id": "8",
              "text": "a",
            },
            Object {
              "children": Array [],
              "id": "9",
            },
            Object {
              "id": "10",
              "text": "bc",
            },
          ],
          "id": "7",
        },
      ],
      "id": "3",
    },
    Object {
      "id": "12",
      "text": "a",
    },
    Object {
      "id": "13",
      "text": "",
    },
    Object {
      "id": "14",
      "text": "b",
    },
  ],
  "id": "0",
}
`;
